rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,201 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <cstdio>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rosbag2_compression/zstd_decompressor.hpp"
+
+#include "rosbag2_storage/filesystem_helper.hpp"
+
+#include "logging.hpp"
+
+namespace
+{
+constexpr const char DECOMPRESSION_IDENTIFIER[] = "zstd";
+
+FILE * open_file(const std::string & uri, const std::string & read_mode)
+{
+  FILE * fp{nullptr};
+#ifdef _WIN32
+  fopen_s(&fp, uri.c_str(), read_mode.c_str());
+#else
+  fp = std::fopen(uri.c_str(), read_mode.c_str());
+#endif
+  return fp;
+}
+
+std::vector<uint8_t> get_input_buffer(
+  const std::string & uri,
+  size_t & compressed_buffer_length)
+{
+  // Get the file size
+  compressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
+
+  // Read in buffer, handling accordingly
+  const auto file_pointer = open_file(uri.c_str(), "rb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary reading!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  // Allocate and read in
+  std::vector<uint8_t> compressed_buffer;
+  compressed_buffer.reserve(compressed_buffer_length);
```

Comment:
You must `.resize()` before using `.data()`
```suggestion
  compressed_buffer.resize(compressed_buffer_length);
```

Reply:
`reserve` was chosen over `resize` due to `resize` default initializing each element.
I'll look into how the compiler chooses to handle this

Reply:
Okay, I understand the performance concern, but using `.data()` after `.reserve()` is not valid: https://stackoverflow.com/questions/38972443/use-stdvectordata-after-reserve

Reply:
Actually, forget `.resize()`, just construct the `vector` with the fill constructor (type (4) shown [here](https://en.cppreference.com/w/cpp/container/vector/vector)).

Please don't be concerned with the penalty of zero-initialization of the elements. That's the cost of using C++ over using C. You can get around it by writing your own allocator that does default-initialization instead of zero-/value-initialization for `uint8_t` and other POD types, but definitely please don't go down that route.

Reply:
Went with`resize`. 
I tried to invoke type (4) via uniform initialization, however the compiler kept selecting the initializer list constructor.

Reply:
Hmm, that's strange. Have you also tried constructing using brackets `()` instead of braces `{}`?

Reply:
That does work, however I think its kind of confusing that old style initialization has to be used.

Reply:
You'll have to come to accept that C++ is nowhere near being a consistent language :)

Reply:
I changed it to use the type (4) constructor and added comments stating that uniform initialization cannot be used for initializing a `std::vector` with size.